src/noit_metric_tag_search.c

```diff
@@ -666,9 +670,13 @@ void
 noit_metric_tag_search_set_arg(noit_metric_tag_search_ast_t *node, int idx, noit_metric_tag_search_ast_t *r) {
   assert(node->operation != OP_MATCH);
   assert(node->operation != OP_NOT_ARGS || idx == 0);
-  if(idx == node->contents.args.cnt) noit_metric_tag_search_resize_args(node, idx+1);
-  assert(node->contents.args.cnt > idx);
+  noit_metric_tag_search_resize_args(node, node->contents.args.nallocd*2);
+  assert(node->contents.args.nallocd > idx);
+  noit_metric_tag_search_ast_t *prev = node->contents.args.node[idx];
   node->contents.args.node[idx] = r;
+  assert(node->contents.args.cnt >= idx); // can't insert with gaps.
+  if(node->contents.args.cnt == idx) node->contents.args.cnt++;
+  else noit_metric_tag_search_free(prev);
```

Comment:
What happens to prev if node->contents.args.cnt == idx ? It looks like it will leak memory

Reply:
If cnt == idx we're appending to the end of an existing list (and not replacing an existing item).  The only idx's that have items in the are less than cnt.